lightning/src/offers/merkle.rs

```diff
@@ -24,6 +25,41 @@ tlv_stream!(SignatureTlvStream, SignatureTlvStreamRef, SIGNATURE_TYPES, {
 	(240, signature: Signature),
 });
 
+/// A function for signing a [`Message`] digest produced from the given [`TaggedHash`]. See
+/// [`TaggedHash`] for details.
+///
+/// The second parameter is either the payer or offer metadata, depending on the TLV stream
+/// represented by the first parameter, and may be useful in order to derive the signing keys.
+pub trait SignFunction<E>: FnOnce(&TaggedHash, &[u8]) -> Result<Signature, E> {}
+
+impl<F, E> SignFunction<E> for F where F: FnOnce(&TaggedHash, &[u8]) -> Result<Signature, E> {}
+
+/// A hash for use in a specific context by tweaking with a context-dependent tag as per [BIP 340]
+/// and computed over the merkle root of a TLV stream to sign as defined in [BOLT 12].
+///
+/// The hash is computed lazily from the TLV stream.
+///
+/// [BIP 340]: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki
+/// [BOLT 12]: https://github.com/rustyrussell/lightning-rfc/blob/guilt/offers/12-offer-encoding.md#signature-calculation
+/// Bytes associated with a tag, which are used to produced a [`Message`] digest to sign.
+pub struct TaggedHash<'a> {
+	tag: &'a str,
+	tlv_stream: &'a [u8],
```

Comment:
I thought signers would need access to the TLV stream's fields and the tag, to verify the actual invoice request/invoice?

Reply:
Yeah, this was an oversight. Exposed as `pub` now.